### ISSUE TITLE
ActiveDirectory users can not login anymore

### ISSUE THREAD

**Issue by moisei** [kevinpapst/kimai2#1569]
Some users always have an error upon login: 

```
Something is wrong

A critical error occurred, please try again. You may have found a software problem, please contact your administrator if the problem persists. ` 
ad browser is redirected to http://my-kimai/en/login_check
```

**To Reproduce**
browse to kimai login page and try to login

**Logfile**
```

[2020-03-18 08:24:40] request.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\UniqueConstraintViolationException: "An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token, auth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params ["Jdoe", "jdoe", "Jdoe@mycompany.com", "jdoe@mycompany.com", 1, null, "", "2020-03-18 08:24:40", null, null, "a:0:{}", "Jonathan doe", "2020-03-18 08:24:40", null, null, null, "ldap"]:  SQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: kimai2_users.email" at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php line 33 {"exception":"[object] (Doctrine\\DBAL\\Exception\\UniqueConstraintViolationException(code: 0): An exception occurred while executing 'INSERT INTO kimai2_users (username, username_canonical, email, email_canonical, enabled, salt, password, last_login, confirmation_token, password_requested_at, roles, alias, registration_date, title, avatar, api_token, auth) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)' with params [\"Jdoe\", \"jdoe\", \"Jdoe@mycompany.com\", \"jdoe@mycompany.com\", 1, null, \"\", \"2020-03-18 08:24:40\", null, null, \"a:0:{}\", \"Jonathan doe\", \"2020-03-18 08:24:40\", null, null, null, \"ldap\"
]:\n\nSQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: kimai2_users.email at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php:33, Doctrine\\DBAL\\Driver\\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: kimai2_users.email at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:123, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 19 UNIQUE constraint failed: kimai2_users.email at /opt/kimai/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:121)"} []

```

**Screenshots**
attached

**Desktop**
 - Device: Windows desktop
 - OS: Windows 10
 - Browser: Chrome
 - Version: latest

**Additional context**
I am using docker-compose from the https://github.com/tobybatch/kimai2.
I've installed the original version 1.5 first and I configured ActiveDirectory integration.
Users were using this instance for a couple of days. There are about 100 users registered.

Then I've changed compose file to use **kimai/kimai2:fpm-alpine-1.8-prod** image and I've upgraded the site accorsing to instructions. 
Also I've installed two plugins **CustomFields** and **Audittrail** and I've created one custom field.
 - Kimai  version: **1.18**
 - PHP version: I don't know, it is coming with the container.

**Note:** the users that have error now, were able to login properly before the upgrade.
Please advise, as it is becoming really serious issue.

<img width="721" alt="2020-03-18 21_17_51-500 - Error" src="https://user-images.githubusercontent.com/1199723/76998817-5a3cd580-695e-11ea-94bf-4c75fdb11ba0.png">

**Comment by kevinpapst** [comment]
The error says that Kimai can't detect the users in the database and tries to re-create them.

Why that happens: I don't know. Did you change your local.yaml / AD configuration?
You did re-install the zend ldap packages after updating Kimai?

What does `git status` say? Maybe your update was 100% successful, if there happened to be a merge conflict while pulling the latest version (in composer.lock at least).

Second question - not directly related, but still an issue: why do you use SQLite and not MySQL? 
That is not supported in production and could lead to data loss upon updates.

**Comment by moisei** [comment]
I've got more information on this error. Hope it can help somehow. The problem seems to be a consequence of two different issues. 
1. some users won't be able to login after the upgrade. Only few users out of more than 100. For everybody else everything is working just fine. I didn't change local.yaml or whatsoever.  My guess that it is somehow related to the password of these users. I know a password of one of them and it has ! at the end. 
2. I tried to delete the user , so it would be recreated upon next login but apparently there is another bug! the user is not fully deleted from the kimai2_user_preferences and kimai2_users_teams tables! so any new user couldn't be created because of constraint violation!
Following queries 
`
select * from kimai2_user_preferences where user_id not in (select id from kimai2_users)
select * from kimai2_users_teams where user_id not in (select id from kimai2_users)
`
returned me bad IDs, which I've just deleted manually and voila! new users can login!
Also, I've changed the password of the problematic user to something simple and he was able to login as well - new user was created properly.

Hope it helps.

**Comment by moisei** [comment]
@kevinpapst  re: Second question - not directly related, but still an issue: why do you use SQLite and not MySQL? That is not supported in production and could lead to data loss upon updates.

I just used the docker compose from the kimai-docker repo and I've assumed it is using mysql. compose file contains and executes mysql service. I was suprised myself by seeng it uses sqlite in fact. Is there a way to migrate to mysql withotu data lost, I would appreciate  any references and also the way to configure kima2 to use mysql from docker-compose.

Pleas note kimai2 is very new thing to me. 3 days ago I've seen it first time.

**Comment by kevinpapst** [comment]
At least the troubles you are seeing when deleting the user come from SQLite. SQLite is a nightmare when it comes to foreign keys and cascade deletes... and therefor not supported in production. 
Kimai uses auto increment fields, if those don't work properly as well (after deleting an entry), then that is a bug in SQLite and not Kimai. For reference: https://www.kimai.org/documentation/installation.html#sqlite-not-recommended-for-production-usage

Maybe you have been lucky that time because your database is so new, but: when deleting users from the database, you will erase all their timesheets as well + preferences + comments + teams ... not a good idea! Disabling them and then changing their username + email in the database might be the better approach.

And I guess that empty foreign keys on the timesheet table (if they were not deleted as well) will lead to several other problems, as Kimai expects the user to be existing on these records.

I don't know how a `!` in the password should affect the login process. The password is passen to the AD bind method. You can try to debug that in https://github.com/kevinpapst/kimai2/blob/master/src/Ldap/LdapManager.php#L94
You could also switch to `APP_ENV=dev` to get more query logs, maybe that helps in finding the reason for the login problem.

Regarding the migration: read #1524 and discuss there please.

For the docker mysql setup: please ask in the docker repository.

**Comment by moisei** [comment]
thank you.  checking now.  I've just hit another issue: waitForDB in startup.sh is not working correct if mysql password contains @ sign.

`
function waitForDB() {
  # Parse sql connection data
  # todo: port is not used atm
  DB_TYPE=$(awk -F '[/:@]' '{print $1}' <<< $DATABASE_URL)
  DB_USER=$(awk -F '[/:@]' '{print $4}' <<< $DATABASE_URL)
  DB_PASS=$(awk -F '[/:@]' '{print $5}' <<< $DATABASE_URL)
  DB_HOST=$(awk -F '[/:@]' '{print $6}' <<< $DATABASE_URL)
  DB_BASE=$(awk -F '[/:@]' '{print $7}' <<< $DATABASE_URL)
`

**Comment by kevinpapst** [comment]
Please share some feedback if you find solution, to help the next user in a similar situation.

**Comment by moisei** [comment]
@kevinpapst I apologize I didn't say a great thanks for this software. Due to the COV-19 crisis we've found all our team almost 150 people is working from home and we urgently need time tracking software. It took me less than a day to put kimai2 to production including ActiveDirectory integration and CustomFields plugin adotation. People already use it successfully without any major obstacles. 

Specifically for this issue, I've provided the queries that helped me to overcome. Now, I am following your suggestion and switching to MySQL.
